WordPress/wordpress-playground
# PHP Blueprints demo page

## What is this PR doing?

Supersedes https://github.com/WordPress/wordpress-playground/pull/1051

Adds a PHP Blueprints demo page where the use of [blueprints.phar](https://github.com/WordPress/blueprints/pull/28) PHP library in Playground may be further explored. The showcase is intentionally not added to http://localhost:5400/website-server/demos/index.html as PHP Blueprints may become a part of Playground core soon enough.

For more context see:

* https://github.com/WordPress/wordpress-playground/issues/1025
* https://github.com/WordPress/blueprints

## How does it work?

* The built Blueprints library is included with this PR via the `blueprints.phar` file
* A number of PHP.wasm improvements have been merged to support it:
   * https://github.com/WordPress/wordpress-playground/pull/1064
   * https://github.com/WordPress/wordpress-playground/pull/1065
   * https://github.com/WordPress/wordpress-playground/pull/1069
* This PR ships a `fetch` subprocess handler to enable streaming network data in the Blueprints library – it uses [a special network transport called `fetch`](https://github.com/WordPress/blueprints/blob/efa8deef56095bd5bcb94868787e29f7b54350f3/src/WordPress/DataSource/PlaygroundFetchSource.php) that requests network data via `proc_open()` when running in Playground. Why subprocesses? They provide everything a custom network handler needs: pipes, asynchronous execution, exit codes, internal PHP plumbing.

## Follow-up work

* Support a real-time progress bar

## Testing instructions

Go to http://localhost:5400/website-server/demos/php-blueprints.html and confirm it looks like on the screenshot below:

![CleanShot 2024-02-28 at 15 46 14@2x](https://github.com/WordPress/wordpress-playground/assets/205419/47a91d99-07f3-40a5-a046-b58f8cda952e)
